### PR TITLE
added split methods to polyline

### PR DESCRIPTION
**Pull request by nmaslarinos**
Added `split` and `split_by_length` methods on `compas.geometry.Polyline` to complement the existing divide methods. These methods return a list of Polylines, similar to the split option in Rhino when dividing a curve. Also, updated the docstrings of the divide methods with examples 😊

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by nmaslarinos** [comment]
I have also developed a `break_polyline_at_point` method that opens a polyline on a specified point if closed or splits it in two. I could add it to this pull request since it's all about Polyline methods, or create a new one

File: src/compas/geometry/primitives/polyline.py [hunk 1]
```diff
@@ -352,14 +352,21 @@ def divide_polyline(self, num_segments):
         list
             list[:class:`~compas.geometry.Point`]
 
+        Examples
+        --------
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide(3)
+        >>> divided_polyline
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/polyline.py:351]: this variable doesn't seem to be defined

File: src/compas/geometry/primitives/polyline.py [hunk 2]
```diff
@@ -379,6 +386,18 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
         The points of the new polyline are constrained to the segments of the old polyline.
         However, since the old points are not part of the new set of points, the geometry of the polyline will change.
 
+        Examples
+        --------
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide_by_length(3)
+        >>> divided_polyline
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/polyline.py:385]: same here?

File: src/compas/geometry/primitives/polyline.py [hunk 3]
```diff
@@ -379,6 +386,18 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
...
+        [Point(0.000, 0.000, 0.000), Point(1.709, 2.418, 0.000), Point(4.051, 3.898, 0.000)]
+
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide_by_length(3, strict=False)
+        >>> divided_polyline
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/polyline.py:390]: and here...

File: src/compas/geometry/primitives/polyline.py [hunk 4]
```diff
@@ -412,6 +431,98 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
 
     divide_by_length = divide_polyline_by_length
 
+    def split_polyline_by_length(self, length, strict=True):
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/polyline.py:426]: what is the difference between split and divide?

File: tests/compas/geometry/test_polyline.py [hunk 5]
```diff
@@ -196,6 +196,140 @@ def test_polyline_split_at_corners(coords, input, expected):
     assert expected == Polyline(coords).split_at_corners(input)
 
 
+@pytest.mark.parametrize(
```
Comment (tomvanmele) [inline tests/compas/geometry/test_polyline.py:186]: could you add a few boundary cases, and tests that are expected to fail?

File: src/compas/geometry/primitives/polyline.py (hunk 3 above)
Comment (nmaslarinos) [reply src/compas/geometry/primitives/polyline.py:390]: Sorry about those, I only now realise that for some reason `invoke testdocs` does not perform a check of the files in src, but of the ones in tests

File: src/compas/geometry/primitives/polyline.py (hunk 4 above)
Comment (nmaslarinos) [reply src/compas/geometry/primitives/polyline.py:426]: The difference is that divide returns a list of the points that the polyline is divided into. Split returns a list of Polylines. It breaks the polyline into different parts, without disregarding it's geometry.

**Comment by tomvanmele** [comment]
if i understand correctly, the division points are placed based on the length of the original polyline. does this mean the length of the segments of the new polyline is somewhat random?

perhaps this should be clearly stated in the docs, because users might be expecting to get a polyline with segments of the specified length as the result...

**Comment by tomvanmele** [comment]
> if i understand correctly, the division points are placed based on the length of the original polyline. does this mean the length of the segments of the new polyline is somewhat random?
> 
> perhaps this should be clearly stated in the docs, because users might be expecting to get a polyline with segments of the specified length as the result...

ha no sorry, i see now that i indeed misunderstood

**Comment by tomvanmele** [comment]
LGTM
if you resolve the merge conflicts and fix the docstrings this is good to go...

**Comment by nmaslarinos** [comment]
I added some boundary cases tests in the `test_polyline`, fixed the docstrings which should now work, and added some additional test to the input length (to accept only values larger than zero and smaller than the length of the polyline)

**Comment by tomvanmele** [comment]
@nmaslarinos almost there :) your test file has some formatting issues (https://github.com/compas-dev/compas/actions/runs/5297432096/jobs/9589200519?pr=1152). other than that it is good to go...

**Comment by nmaslarinos** [comment]
Seems like after formating with black, there is still an [error on the mac-os 3.7 build](https://github.com/compas-dev/compas/actions/runs/5309441021/jobs/9610099589) __FAILED tests/compas/datastructures/test_graph.py::test_graph_networkx_conversion - ModuleNotFoundError: No module named '_bz2'__ . Any idea what this could be?

**Comment by tomvanmele** [comment]
no worries this is a github hickup

**Comment by tomvanmele** [comment]
i will rerun and the problem probably goes away (had the same a few days ago)

**Comment by nmaslarinos** [comment]
I tried to do that, but for the love of me I couldn't find the rerun button 🙄

**Review by tomvanmele — APPROVED** [review]
